x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewService.java

```diff
@@ -0,0 +1,321 @@
+        Setting.Property.Dynamic
+    );
+    public static final Setting<Integer> MAX_VIEW_DEPTH_SETTING = Setting.intSetting(
+        "esql.views.max_view_depth",
+        10,
+        0,
+        100,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private volatile int maxViewsCount;
+    private volatile int maxViewLength;
+    // TODO: not yet used, but will be
+    private volatile int maxViewDepth;
+
+    public ViewService(ClusterService clusterService, ProjectResolver projectResolver, Settings settings) {
+        this.clusterService = clusterService;
+        this.projectResolver = projectResolver;
+        this.taskQueue = clusterService.createTaskQueue(
+            "update-esql-view-metadata",
+            Priority.NORMAL,
+            new SequentialAckingBatchedTaskExecutor<>()
+        );
+        this.telemetry = new PlanTelemetry(new EsqlFunctionRegistry());
+        this.maxViewsCount = MAX_VIEWS_COUNT_SETTING.get(settings);
+        this.maxViewLength = MAX_VIEW_LENGTH_SETTING.get(settings);
+        this.maxViewDepth = MAX_VIEW_DEPTH_SETTING.get(settings);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(MAX_VIEWS_COUNT_SETTING, (i) -> this.maxViewsCount = i);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(MAX_VIEW_LENGTH_SETTING, (i) -> this.maxViewLength = i);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(MAX_VIEW_DEPTH_SETTING, (i) -> this.maxViewDepth = i);
+    }
+
+    ViewMetadata getMetadata() {
+        return getMetadata(projectResolver.getProjectId());
+    }
+
+    protected ViewMetadata getMetadata(ProjectMetadata projectMetadata) {
+        return projectMetadata.custom(ViewMetadata.TYPE, ViewMetadata.EMPTY);
+    }
+
+    ViewMetadata getMetadata(ProjectId projectId) {
+        return getMetadata(clusterService.state().metadata().getProject(projectId));
+    }
+
+    /**
+     * Adds or modifies a view by name.
+     */
+    public void putView(ProjectId projectId, PutViewAction.Request request, ActionListener<AcknowledgedResponse> listener) {
+        if (viewsFeatureEnabled() == false) {
+            listener.onFailure(new IllegalArgumentException("ESQL views are not enabled"));
+            return;
+        }
+
+        final View view = request.view();
+        final ProjectMetadata metadata = clusterService.state().metadata().getProject(projectId);
+        try {
+            validatePutView(metadata, view);
+        } catch (Exception e) {
+            listener.onFailure(e);
+            return;
+        }
+        final AckedClusterStateUpdateTask task = new AckedClusterStateUpdateTask(request, listener) {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                final ProjectMetadata project = currentState.metadata().getProject(projectId);
+                final ViewMetadata viewMetadata = project.custom(ViewMetadata.TYPE, ViewMetadata.EMPTY);
```

Comment:
Might as well make use of the helpers we add :D
```suggestion
                final ViewMetadata viewMetadata = getMetadata(project);
```